### ISSUE TITLE
Menu item dependencies (dependsOnModule, dependsOnConfig) are broken

### ISSUE THREAD

**Issue by mam08ixo** [magento/magento2#9720]
### Preconditions
1. Magento 2.2.0-dev is installed (bdb04648ab at the time of writing)
2. Any `menu.xml` adds menu items with dependency to
  - module that is not installed
  - config value that evaluates to false

```xml
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Backend:etc/menu.xsd">
    <menu>
        <add id="Foo_MenuItemDependencies::dependencies"
             title="Menu Item Dependencies"
             translate="title"
             module="Foo_MenuItemDependencies"
             sortOrder="10"
             parent="Magento_Backend::system"
             resource="Foo_MenuItemDependencies::dependencies"
        />

        <add id="Foo_MenuItemDependencies::module_dependency"
             title="Depends On Module"
             module="Foo_MenuItemDependencies"
             sortOrder="10"
             resource="Foo_MenuItemDependencies::dependencies"
             parent="Foo_MenuItemDependencies::dependencies"
             action="foo/module"
             dependsOnModule="Foo_Bar"
        />

        <add id="Foo_MenuItemDependencies::config_dependency"
             title="Depends On Config"
             module="Foo_MenuItemDependencies"
             sortOrder="20"
             resource="Foo_MenuItemDependencies::dependencies"
             parent="Foo_MenuItemDependencies::dependencies"
             action="foo/config"
             dependsOnConfig="foo/config/enabled"
        />
    </menu>
</config>

```

### Steps to reproduce
1. Log in to admin panel
2. Open up the top-level menu item that contains the menu items in question

### Expected result
The menu items are not visible.

### Actual result
The menu items are visible.

### Additional information
The menu item handling was [refactored](https://github.com/magento/magento2/commit/24b608a990cefa0a0f5421c3ec541c44bf3259b4) to [read item data from two different sources](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/Model/Menu/Item.php#L488):
* from original XML definition if cache is empty
* from [transformed item data](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/Model/Menu/Item.php#L463) when available in cache

The issue is that the initial reading from the uncached `menu.xml` node fails because the array keys used in `populateFromArray` do not conform to the XSD. That in turn means that various attributes are never taken over to the transformed/cached menu entry and the dependency checks always [return](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/Model/Menu/Item.php#L428) [true](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/Model/Menu/Item.php#L441).

The [corresponding unit test](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/Test/Unit/Model/Menu/ItemTest.php#L265) does not reveal this issue because it makes wrong/incomplete assumptions on the [incoming data](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/Test/Unit/Model/_files/menu_item_constructor_data.php). So yes, the code in question is covered but not the scenario where menu item data comes in as-is from the `menu.xml` file.

Another observation I made is that, although the same problem should apply to the `sortOrder` attribute, sorting actual works. That is because this attribute gets evaluated [in another class](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/Model/Menu/Builder.php#L70) where the original `menu.xml` contents are available.

Quick fix would probably be to align the [array keys used during transformation](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/Model/Menu/Item.php#L463) with the [attribute names in the XSD](https://github.com/magento/magento2/blob/bdb04648ab33cf74a60cf20915e3eda4ba7dca66/app/code/Magento/Backend/etc/menu.xsd#L22).

**Comment by magento-engcom-team** [comment]
@mam08ixo, thank you for your report.
The issue is already fixed in develop branch
But we will consider to backport the fix to patch releases

**Comment by hannassy** [comment]
#mageconf

**Comment by okorshenko** [comment]
Hi @mam08ixo 
The issue has been fixed and delivered in 2.2-develop branch. Will be available with upcoming patch release.
Fixed in magento/magento2#12747 by @hannassy in 2.2-develop branch
   Related commits:
     1. ddf6c9f
